extensions/amp-story/1.0/amp-story-quiz.js

```diff
@@ -299,13 +363,158 @@ export class AmpStoryQuiz extends AMP.BaseElement {
+   */
+  updateReactionData_(reactionValue) {
+    this.executeReactionRequest_(
+      dict({
+        'method': 'POST',
+      }),
+      reactionValue
+    ).catch(error => {
+      dev().error(TAG, error);
+    });
+  }
 
-      this.hasReceivedResponse_ = true;
+  /**
+   * Executes a Reactions API call.
+   *
+   * @param {Object} requestOptions
+   * @param {number} reactionValue
+   * @return {Promise<JsonObject>}
+   * @private
+   */
+  executeReactionRequest_(requestOptions, reactionValue) {
+    // TODO(jackbsteinberg): Add a default reactions endpoint.
+    if (!assertAbsoluteHttpOrHttpsUrl(this.element.getAttribute('endpoint'))) {
+      return Promise.reject(ENDPOINT_INVALID_ERROR);
+    }
+
+    if (this.reactionId_ === null) {
+      const quizPageId = closest(dev().assertElement(this.element), el => {
+        return el.tagName.toLowerCase() === 'amp-story-page';
+      }).getAttribute('id');
+
+      this.reactionId_ = `CANONICAL_URL#page=${quizPageId}`;
+    }
+
+    const requestVars = dict({
+      'reactionType': STORY_REACTION_TYPE_QUIZ,
+      'reactionId': this.reactionId_,
     });
+
+    let url = this.element.getAttribute('endpoint');
+
+    if (requestOptions['method'] === 'POST') {
+      requestVars['reactionValue'] = reactionValue;
+      requestOptions['body'] = requestVars;
+    } else if (requestOptions['method'] === 'GET') {
+      url = addParamsToUrl(url, requestVars);
+    }
+
+    return this.getClientId_().then(clientId => {
+      requestVars['clientId'] = clientId;
+      return this.requestService_.executeRequest(url, null, requestOptions);
+    });
+  }
+
+  /**
+   * Handles incoming reaction data response
+   *
+   * RESPONSE FORMAT
+   * {
+   *  totalResponseCount: <number>
+   *  hasUserResponded: <boolean>
+   *  responses: [
+   *    {
+   *      reactionValue:
+   *      totalCount:
+   *      selectedByUser:
+   *    },
+   *    ...
+   *  ]
+   * }
+   * @param {ReactionResponseType} response
+   * @private
+   */
+  handleSuccessfulDataRetrieval_(response) {
+    if (!(response && 'data' in response)) {
+      return;
+    }
+
+    this.responseData_ = {
+      totalCount: response.data.totalResponseCount,
+      data: response.data.responses,
+    };
+
+    this.hasUserSelection_ = response.data.hasUserResponded;
+    if (this.hasUserSelection_) {
+      const selectedOptionKey = this.responseData_.data.find(
+        response => response.selectedByUser
+      ).reactionValue;
+
+      const options = this.quizEl_.querySelectorAll(
+        '.i-amphtml-story-quiz-option'
+      );
+
+      if (selectedOptionKey >= options.length) {
+        dev().error(
+          TAG,
+          `Quiz #${this.element.getAttribute('id')} does not have option ${
+            answerChoiceOptions[selectedOptionKey]
+          }, but user selected option ${answerChoiceOptions[selectedOptionKey]}`
+        );
+        return;
+      }
+
+      this.quizEl_.classList.add('i-amphtml-story-quiz-post-selection');
```

Comment:
Can we factorize this with the code in `handleOptionSelection_()`? It'd be great if there was only one piece of code that changes the state of the quiz to the "answered" view.

Reply:
I was actually working on this when this suggestion came through, I agree it would be a change in favor of clarity